sdk/lib/drivers/sound/mmebuddy/mmewrap.c

```diff
@@ -365,3 +365,82 @@ MmeGetPosition(
     return Result;
 }
 
+MMRESULT
+MmeGetVolume(
+    _In_ MMDEVICE_TYPE DeviceType,
+    _In_ DWORD DeviceId,
+    _In_ DWORD_PTR PrivateHandle,
+    _Out_ PDWORD_PTR pdwVolume)
+{
+    MMRESULT Result;
+    PSOUND_DEVICE_INSTANCE SoundDeviceInstance;
+    PSOUND_DEVICE SoundDevice;
+    PMMFUNCTION_TABLE FunctionTable;
+
+    /* Sanity check */
+    SND_ASSERT(DeviceType == AUX_DEVICE_TYPE ||
+               DeviceType == MIDI_OUT_DEVICE_TYPE ||
+               DeviceType == WAVE_OUT_DEVICE_TYPE);
+
+    VALIDATE_MMSYS_PARAMETER(PrivateHandle);
+    SoundDeviceInstance = (PSOUND_DEVICE_INSTANCE)PrivateHandle;
+
+    if (!IsValidSoundDeviceInstance(SoundDeviceInstance))
+        return MMSYSERR_INVALHANDLE;
+
+    Result = GetSoundDeviceFromInstance(SoundDeviceInstance, &SoundDevice);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    Result = GetSoundDeviceFunctionTable(SoundDevice, &FunctionTable);
+    if (!MMSUCCESS(Result))
+        return TranslateInternalMmResult(Result);
+
+    if (!FunctionTable->GetVolume)
+        return MMSYSERR_NOTSUPPORTED;
+
+    /* Call the driver */
+    Result = FunctionTable->GetVolume(SoundDeviceInstance, DeviceId, (PDWORD)pdwVolume);
+
+    return Result;
```

Comment:
Here there's no need to call the TranslateInternal helper?

Reply:
No. This helper is generally used only in one failure case handling -  when any of parameters are invalid. It does not change anything in all other cases. All it does is only converting two return values (which are not returned by the calling function) to the general error code. For any other values, it simply returns the same result.
The function I'm calling in this case, has all parameters valid, so using the helper will not change anything.
See its implementation at https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/drivers/sound/mmebuddy/utility.c;hb=f69e256376e939e63029c8d8be8b97eaa206eec0#l132.